### PR TITLE
operator-sdk: run bundle{-upgrade} support insecure registry server

### LINKED ISSUE #4815: run bundle{-upgrade} should able to pull image from insecure registry
## Feature Request

Optional support for `operator-sdk run bundle/bundle-upgrade`  pulling the bundle image from an insecure container registry.

#### Describe the problem you need a feature to resolve.

I run a local insecure docker image for testing. I push the generated operator bundle image to the local docker registry.
Running the `operator-sdk run bundle localhost:5000/pmem-csi-bundle:v0.9.0` fails to install the operator:

```
$ operator-sdk run bundle localhost:5000/pmem-csi-bundle:v0.9.0 --verbose
FATA[0000] Failed to run bundle: pull bundle image: error pulling image localhost:5000/pmem-csi-bundle:v0.9.0: error resolving name : failed to do request: Head "https://localhost:5000/v2/pmem-csi-bundle/manifests/v0.9.0": http: server gave HTTP response to HTTPS client 
```

#### Describe the solution you'd like.

The `opm` tool already supports pulling images from the insecure registery when asked using `--skip-tls`, `operator-sdk` could make used of this to support insecure registries. Provide a command-line argument for `operator-sdk run bundle` to enable/disable the `--skip-tls` :

     `$ operator-sdk run bundle <bundleImage> [--allow-insecrue]`

### PR DESCRIPTION
**Pull request by avalluri**
Supporting insecure registry is useful for testing operator installation
from a bundle image stored at the local registry.

Closes #4815

**Description of the change:**

Add a new boolean command-line option(defaults to false) named `--allow-insecure` for both `operator-sdk run bundle` and `operator-sdk run bundle-upgrade`. When enabled this argument the commands ignores the SSL errors that occurred while communicating the docker registry. This allows pulling the bundle images from insecure(usually local) docker registries.

**Motivation for the change:**

Using a local insecure registry is a possible case for storing the generated bundle image, and test the operator installation.

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: internal/olm/operator/bundle/install.go [hunk 1]
```diff
@@ -51,6 +51,7 @@ func (i *Install) BindFlags(fs *pflag.FlagSet) {
 
 	// --mode is hidden so only users who know what they're doing can alter add mode.
 	fs.StringVar((*string)(&i.BundleAddMode), "mode", "", "mode to use for adding bundle to index")
+	fs.BoolVar(&i.IndexImageCatalogCreator.AllowInsecure, "allow-insecure", false, "allows pulling bundle image from an insecure registry server")
```
Comment (estroz) [inline internal/olm/operator/bundle/install.go:54]: How about mirroring `opm` and naming this
```suggestion
	fs.BoolVar(&i.IndexImageCatalogCreator.SkipTLS, "skip-tls", false, "skip authentication of image registry TLS certificate when pulling a bundle image in-cluster")
```

Comment (avalluri) [reply internal/olm/operator/bundle/install.go:54]: > How about mirroring `opm` and naming this

Sounds good, i will update the PR.

Comment (avalluri) [reply internal/olm/operator/bundle/install.go]: Renamed the option as suggested. Also moved the flag to `IndexImageCatalogCreator` in `internal/olm/operator/registry/index_image.go` so the option will be available to both `run bundle` and `run bundle-upgrade`.

File: internal/olm/operator/registry/index/registry_pod.go [hunk 2]
```diff
@@ -50,6 +50,8 @@ type BundleItem struct {
 	ImageTag string `json:"imageTag"`
 	// AddMode describes how the bundle should be added to an index image.
 	AddMode BundleAddMode `json:"mode"`
+	// SkipTLS controls wether to ignore SSL errors while pulling bundle image from registry server.
+	SkipTLS bool `json:"SkipTLS"`
```
Comment (estroz) [inline internal/olm/operator/registry/index/registry_pod.go:54]: This field should be added to `RegistryPod` not `BundleItem` since it's globally configured and not on a per-bundle basis.

Comment (jberkhahn) [reply internal/olm/operator/registry/index/registry_pod.go:54]: that's also where the preexisting cert stuff is anyway :+1:

Comment (jmrodri) [reply internal/olm/operator/registry/index/registry_pod.go:54]: agreed

File: internal/olm/operator/registry/index/registry_pod.go [hunk 3]
```diff
@@ -302,7 +304,7 @@ func newBool(b bool) *bool {
 
 const cmdTemplate = `/bin/mkdir -p {{ dirname .DBPath }} && \
 {{- range $i, $item := .BundleItems }}
-/bin/opm registry add -d {{ $.DBPath }} -b {{ $item.ImageTag }} --mode={{ $item.AddMode }}{{ if $.CASecretName }} --ca-file=/certs/cert.pem{{ end }} && \
+/bin/opm registry add -d {{ $.DBPath }} -b {{ $item.ImageTag }} --mode={{ $item.AddMode }}{{ if $.CASecretName }} --ca-file=/certs/cert.pem{{ end }} --skip-tls={{ $item.SkipTLS }} && \
```
Comment (jmrodri) [inline internal/olm/operator/registry/index/registry_pod.go:307]: :+1:  Great catch. When I was playing around to add this locally I didn't notice having to do this.

File: internal/olm/operator/registry/index/registry_pod_test.go [hunk 4]
```diff
@@ -114,6 +132,11 @@ var _ = Describe("RegistryPod", func() {
 						ImageTag: "quay.io/example/example-operator-bundle:1.0.1",
 						AddMode:  SemverBundleAddMode,
 					},
+					BundleItem{
+						ImageTag: "localhost/example-operator-bundle:1.0.1",
+						AddMode:  SemverBundleAddMode,
+						SkipTLS:  true,
```
Comment (jmrodri) [inline internal/olm/operator/registry/index/registry_pod_test.go:138]: Like was said earlier, this field should be on RegistryPod.

File: internal/olm/operator/registry/index/registry_pod_test.go [hunk 5]
```diff
@@ -91,6 +92,14 @@ var _ = Describe("RegistryPod", func() {
 				Expect(rp.DBPath).To(Equal("/database/index.db"))
 			})
 
+			It("should create a registry with custom database path", func() {
+				customDBPath := "/foo/bar/database.db"
```
Comment (jmrodri) [inline internal/olm/operator/registry/index/registry_pod_test.go:96]: Do we allow a custom database? This seems like we're mixing 2 different items.

File: internal/olm/operator/registry/index/registry_pod_test.go [hunk 6]
```diff
@@ -224,7 +247,7 @@ func containerCommandFor(dbPath string, items []BundleItem, hasCA bool) string {
 	}
 	additions := &strings.Builder{}
 	for _, item := range items {
-		additions.WriteString(fmt.Sprintf("/bin/opm registry add -d %s -b %s --mode=%s%s && \\\n", dbPath, item.ImageTag, item.AddMode, caFlag))
+		additions.WriteString(fmt.Sprintf("/bin/opm registry add -d %s -b %s --mode=%s%s --skip-tls=%v && \\\n", dbPath, item.ImageTag, item.AddMode, caFlag, item.SkipTLS))
 	}
-	return fmt.Sprintf("/bin/mkdir -p /database && \\\n%s/bin/opm registry serve -d /database/index.db -p 50051\n", additions.String())
+	return fmt.Sprintf("/bin/mkdir -p %s && \\\n%s/bin/opm registry serve -d %s -p 50051\n", filepath.Dir(dbPath), additions.String(), dbPath)
```
Comment (jmrodri) [inline internal/olm/operator/registry/index/registry_pod_test.go:252]: I didn't think we were allowing custom database path. I think that should be a separate PR.

**Review by jmrodri — CHANGES_REQUESTED** [review]
All in all a GREAT PR. I have 2 items:
1. the custom DB path should be a separate PR
2. the SkipTLS field should be on RegistryPod like others mentioned.

File: internal/olm/operator/registry/index/registry_pod_test.go (hunk 6 above)
Comment (avalluri) [reply internal/olm/operator/registry/index/registry_pod_test.go:252]: I added this custom database change in this PR as the CI reported a failure:

```
go vet ./...
tools/scripts/fetch golangci-lint 1.31.0 && tools/bin/golangci-lint run
golangci-lint missing or not version '1.31.0', downloading...
golangci/golangci-lint info checking GitHub for tag 'v1.31.0'
golangci/golangci-lint info found version: 1.31.0 for v1.31.0/linux/amd64
golangci/golangci-lint info installed /home/runner/work/operator-sdk/operator-sdk/tools/bin/golangci-lint
internal/olm/operator/registry/index/registry_pod_test.go:234:26: `containerCommandFor` - `dbPath` always receives `defaultDBPath` (`"/database/index.db"`) (unparam)
func containerCommandFor(dbPath string, items []BundleItem, hasCA bool) string {
                         ^
Makefile:122: recipe for target 'test-sanity' failed
make: *** [test-sanity] Error 1
Error: Process completed with exit code 2.
```

**Comment by avalluri** [comment]
> All in all a GREAT PR.

Thanks! for your review time. Sorry for late responce.
 I have 2 items:

> 
> 1. the custom DB path should be a separate PR

I removed this change from this PR. I will try to send it as a separate PR.

> 2. the SkipTLS field should be on RegistryPod like others mentioned.

Changed as suggested.

File: internal/olm/operator/registry/index/registry_pod.go (hunk 2 above)
Comment (avalluri) [reply internal/olm/operator/registry/index/registry_pod.go]: Done. Changed as @estroz suggested.

File: internal/olm/operator/registry/index/registry_pod_test.go (hunk 4 above)
Comment (avalluri) [reply internal/olm/operator/registry/index/registry_pod_test.go]: Done.

**Review by estroz — APPROVED** [review]
/lgtm

File: internal/olm/operator/registry/index/registry_pod_test.go (hunk 5 above)
Comment (avalluri) [reply internal/olm/operator/registry/index/registry_pod_test.go]: > Do we allow a custom database?

I am not sure, in case, we don't support then we could remove `dbPath` parameter from `containerCommandFor` in the unit tests.

Comment (avalluri) [reply internal/olm/operator/registry/index/registry_pod_test.go]: At the same time, I don't see any logical reason for not supporting the custom database path.

**Review by jmrodri — APPROVED** [review]
/lgtm
